src/realm/util/file.cpp

```diff
@@ -689,38 +689,86 @@ void File::prealloc(size_t size)
     prealloc_if_supported(0, size);
 
 #else // Non-atomic fallback
-    if (size <= to_size_t(get_size())) {
+
+#if REALM_PLATFORM_APPLE
+    // posix_fallocate() is not supported on MacOS or iOS, so use a combination of fcntl(F_PREALLOCATE) and ftruncate().
+
+    // the following block is equivilent to this->get_size() but done manually so we can reuse the fstat call
+    // for querying the st_blocks later on.
+    SizeType cur_file_size;
+    struct stat statbuf;
+    { // get_size()
```

Comment:
Duplicating this logic is kinda ugly, and really clutters up the logic of what the `preallocate` function is trying to do. It also requires that we duplicate the weird logic related to encryption key size. Extracting a function that returned both the logical and physical file size would be clearer, but it may be hard to cope with Windows where we don't care about the physical file size and it'd incur extra expense to calculate.

Reply:
The extra expense is not really an issue, this code is not on a critical path.

Reply:
The ugliness is :-)